### DISCUSSION TITLE
Extract rect of Drawings

### DISCUSSION THREAD

**Discussion by rafayaar** [pymupdf/PyMuPDF#4065] [General]
Hi @JorjMcKie 
I am back again with another question.

I am trying to extract drawings from PDF as images. I aiming to achieve that each drawing in PDF is saved as an image with me. Along with that I get their rects as well. This will be me render it back to the PDF and achieve same PDF.

I create a new page on empty PDF and one by one copy drawings and get_pixmap with alpha as true which gives me a transparent background image with drawing on it.
For each page I create new page with same dimensions of page.
I am looking to achieve the dimensions which exactly fit the drawings

```python
import fitz 

doc = fitz.open("PDFs/dinnerly.pdf")
page = doc[0]
paths = page.get_drawings()  

output_dir = "pdf-extracted-drawings/"

import os
os.makedirs(output_dir, exist_ok=True)

for idx, path in enumerate(paths):
    outpdf = fitz.open()
    outpage = outpdf.new_page(width=page.rect.width, height=page.rect.height)
    shape = outpage.new_shape()

    for item in path["items"]:
        if item[0] == "l":  # Line
            shape.draw_line(item[1], item[2])
        elif item[0] == "re":  # Rectangle
            shape.draw_rect(item[1])
        elif item[0] == "qu":  # Quad
            shape.draw_quad(item[1])
        elif item[0] == "c":  # Bezier curve
            shape.draw_bezier(item[1], item[2], item[3], item[4])
        else:
            raise ValueError("Unhandled drawing", item)

    shape.finish(
        fill=path["fill"],  # Fill color
        color=path["color"],  # Line color
        dashes=path["dashes"],  # Line dashing
        even_odd=path.get("even_odd", True),  # Control color of overlaps
        closePath=path["closePath"],  # Whether to connect last and first point
        width=path["width"],  # Line width
    )
    shape.commit()

    pix = outpage.get_pixmap(alpha=True)
    image_path = os.path.join(output_dir, f"drawing_{idx + 1}.png")
    pix.save(image_path)  # Save as PNG
    print(f"Saved drawing {idx + 1} to {image_path}")

``` 

How can I extract exact dimension of Draw so I can create new page based on that and get exact fit

**Comment by rafayaar** [comment]
Also, I would want to know how to manage z-index such that image drawing is not overlapped randomly but based on z-index extraction of each drawing.
Also for images as well

**Comment by JorjMcKie** [comment]
There is a utility `Page` method `cluster_drawings()`, https://pymupdf.readthedocs.io/en/latest/page.html#Page.cluster_drawings, which goes through vector graphics and joins the boundary boxes of any that are close enough to each other.
You can use the returned list of rectangles as clips for making appropriate pixmaps for your purpose.

Otherwise:
Using `Shape` to recreate drawings on a new page works quite well, but it is not perfect (and will never be). E.g. there is no full support of alternate coloring using non-zero-winding rule and other details.

**Reply by rafayaar** [reply]
`cluster_drawings` only respect Rects. There is line, quad, bezier curve as well.

**Reply by JorjMcKie** [reply]
You misunderstand the purpose of that method:
It delivers the rectangle hull of interconnected vector graphics. For purposes like using this as the clip for making pixmaps.
